WC2-110 workflows READ

beygorghor opened the pull request:
List/Details view of workflows version for a specific beneficiary type.

## Self proof reading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] Did I add translations
- [ ] My migrations file are included
- [ ] Are there enough tests

## Changes

New workflows domain only to READ


## How to test

Go to the `/admin/iaso/workflow/`
Create a new workflow linked to a entity type.
Go to the `/admin/iaso/workflowversion/`
Create multiple versions of the workflow you created.
You can add changes and followups too.
For Changes: Mapping should be something like 
`{"key":"value"}`
For Followups: condition is a jsonlogic expression (true is also working)

Test filters, sort, results, pagination, navigation


## Print screen / video
<img width="847" alt="Screenshot 2022-12-12 at 14 28 07" src="https://user-images.githubusercontent.com/12494624/207057786-77522192-03ec-4119-b44b-2a72062cc488.png">
<img width="831" alt="Screenshot 2022-12-12 at 14 32 14" src="https://user-images.githubusercontent.com/12494624/207057829-f6a7ff39-ed40-4d18-8433-e3e022d82b84.png">
<img width="1677" alt="Screenshot 2022-12-12 at 14 33 52" src="https://user-images.githubusercontent.com/12494624/207057849-2180732c-3c93-46d1-8b67-0f46a29884ea.png">

beygorghor committed:
Adding workflows list page

beygorghor committed:
Merge branch 'main' into WC2-110-WORKFLOWS

beygorghor committed:
change icon + add name column

beygorghor committed:
adding filters

beygorghor committed:
workflow details wip

beygorghor committed:
follow up and changes

beygorghor committed:
apply demo changes

beygorghor committed:
apply demo changes

beygorghor committed:
Merge branch 'WC2-82-workflow-management' into WC2-110-WORKFLOWS

beygorghor committed:
map to new api

beygorghor committed:
Merge branch 'WC2-82-workflow-management' into WC2-110-WORKFLOWS

beygorghor committed:
Merge branch 'WC2-82-workflow-management' into WC2-110-WORKFLOWS

beygorghor committed:
Plug the api

beygorghor committed:
Merge branch 'WC2-82-workflow-management' into WC2-110-WORKFLOWS

beygorghor committed:
filters

beygorghor committed:
Merge branch 'WC2-82-workflow-management' into WC2-110-WORKFLOWS

beygorghor committed:
hide action columns

beygorghor committed:
Merge branch 'WC2-82-workflow-management' into WC2-110-WORKFLOWS

quang-le commented on hat/assets/js/apps/Iaso/constants/routes.js:
```diff
@@ -790,12 +792,49 @@ export const storageDetailPath = {
+    baseUrl: baseUrls.workflows,
+    permissions: ['iaso_workflows'],
+    component: props => <Workflows {...props} />,
+    params: [
```
Add unrequired accountId param in anticipation of the accountId PR

quang-le commented on hat/assets/js/apps/Iaso/constants/routes.js:
```diff
@@ -790,12 +792,49 @@ export const storageDetailPath = {
+    baseUrl: baseUrls.workflowDetail,
+    permissions: ['iaso_workflows'],
+    component: props => <WorkflowDetails {...props} />,
+    params: [
```
accountId cf supra

quang-le commented on hat/assets/js/apps/Iaso/domains/forms/components/LinkToForm.tsx:
```diff
@@ -1,19 +1,33 @@
+}));
 export const LinkToForm: FunctionComponent<Props> = ({ formId, formName }) => {
+    const classes: Record<string, string> = useStyles();
     const user = useCurrentUser();
     if (userHasPermission('iaso_forms', user)) {
         const formUrl = `/${baseUrls.formDetail}/formId/${formId}`;
-        return <Link to={formUrl}>{formName}</Link>;
+        return (
+            <Link className={classes.link} to={formUrl}>
+                {formName}
+            </Link>
+        );
     }
     return <>{formName}</>;
 };
```
This should probably be made into a generic component for all links that appear in tables (at least), eg: users (emails) and polio grouped campaigns

beygorghor replied:
This is a component like `LinkToOrgUnit`, how do you want to make it more generic? it's in form domain

quang-le commented on hat/assets/js/apps/Iaso/domains/forms/components/LinkToForm.tsx:
```diff
@@ -1,19 +1,33 @@
+
+const useStyles = makeStyles(theme => ({
+    link: {
+        color: theme.palette.primary.main,
```
This will not work well in polio as the primary is pretty close to black 
![Screenshot 2022-12-15 at 11 29 17](https://user-images.githubusercontent.com/38907762/207836211-00a4b88b-9b6f-46c6-aacd-cf6a5385428f.png)

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/config.tsx:
```diff
@@ -0,0 +1,190 @@
+                return (
+                    <>
+                        {Object.keys(settings.row.original.mapping)
+                            .map(mapKey => mapKey)
```
shouldn't we show what the key is mapped to?

beygorghor replied:
it was not like that on the mock-ups

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/hooks/requests/useGetWorkflowVersions.ts:
```diff
@@ -0,0 +1,64 @@
+const getWorkflowVersions = async (
+    options: WorkflowsParams,
+): Promise<WorkflowVersionsPaginated> => {
+    const { pageSize, entityTypeId, ...params } = options as Record<
```
if accountId is added to params, it should be removed here in order not to be passed to the API

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/hooks/requests/useGetWorkflowVersions.ts:
```diff
@@ -0,0 +1,64 @@
+};
+
+export const useGetWorkflowVersions = (
+    options: WorkflowsParams,
```
cf supra

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/index.tsx:
```diff
@@ -0,0 +1,91 @@
+const useStyles = makeStyles(theme => ({
+    ...commonStyles(theme),
+}));
+// TODO: centralize those types
```
yes 😀

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/index.tsx:
```diff
@@ -0,0 +1,91 @@
+                    } else {
+                        dispatch(redirectToReplace(baseUrls.entityTypes, {}));
+                    }
+                }}
```
We could make a hook for this. It would save us the import and declaration of `dispatch` and `prevPathName`

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/index.tsx:
```diff
@@ -0,0 +1,91 @@
+            entityType
+                ? `${entityType.name} - ${formatMessage(MESSAGES.workflows)}`
+                : '',
+        [entityType, formatMessage],
```
You don't need to memoïze strings

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/index.tsx:
```diff
@@ -0,0 +1,91 @@
+                    count={data?.count ?? 0}
+                    params={params}
+                    onTableParamsChange={p =>
+                        dispatch(redirectToReplace(baseUrl, p))
```
But then we need dispatch here... hum...🤔

beygorghor replied:
it makes more sense anyway to group types somewhere

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/config.tsx:
```diff
@@ -0,0 +1,190 @@
+            accessor: 'condition',
+        },
+        {
+            Header: formatMessage(MESSAGES.forms),
```
I think `forms` should be singular

quang-le commented on hat/assets/js/apps/Iaso/routing/redirections.js:
```diff
@@ -63,6 +64,14 @@ const addRoutes = (baseRoutes, overrideLanding) => {
                 storageDefaultSort,
             )}/pageSize/20/page/1`}
         />,
+        <Redirect
+            path={`${baseUrls.workflows}/entityTypeId/:entityTypeId`}
+            to={`${
+                baseUrls.workflows
+            }/entityTypeId/:entityTypeId/order/${getSort(
+                workflowDefaultSort,
+            )}/pageSize/20/page/1`}
+        />,
```
probably need to add accountId here as well

quang-le replied:
This change requires the rest of the accountId PR to work. Changing now will break the page

beygorghor replied:
💥

quang-le reviewed (commented):
Nothing really blocking. Some changes that should be made in anticipation of the accountId PR

quang-le committed:
WC2-110: handle json content in table

quang-le committed:
WC2-110: add optional accountId param

- add param in route
- filter it out params sent to backend

quang-le committed:
WC2-110: small fixes

- update translation for forms
- fix wrong type import
- remove useless memoization

beygorghor committed:
code review

quang-le commented on hat/assets/js/apps/Iaso/routing/useGoBack.tsx:
```diff
@@ -0,0 +1,25 @@
+import { useDispatch, useSelector } from 'react-redux';
+
+import { baseUrls } from '../constants/urls';
+
+import { redirectToReplace } from './actions';
+
+type RouterCustom = {
+    prevPathname: string | undefined;
+};
+type Router = {
+    goBack: () => void;
+};
+type State = {
+    routerCustom: RouterCustom;
+};
+export const useGoBack = (router: Router): (() => void) => {
+    const prevPathname = useSelector(
+        (state: State) => state.routerCustom.prevPathname,
+    );
+    const dispatch = useDispatch();
+    if (prevPathname) {
+        return () => router.goBack();
+    }
+    return () => dispatch(redirectToReplace(baseUrls.entityTypes, {}));
+};
```
Awesome. @codeKameleon check this out, might be useful next time you need to do routing 👍
